FirebaseAuthUI/FUIAuth.m

```diff
@@ -250,6 +283,57 @@ - (void)signInWithProviderUI:(id<FUIAuthProvider>)providerUI
   }];
 }
 
+- (void)signInWithEmailHint:(NSString *)emailHint
+   presentingViewController:(UIViewController *)presentingViewController
+                 completion:(FIRAuthDataResultCallback)completion {
+  NSString *kTempApp = @"tempApp";
+  FIROptions *options = [FIROptions defaultOptions];
+  if (![FIRApp appNamed:kTempApp]) {
+    [FIRApp configureWithName:kTempApp options:options];
+  }
+  FIRApp *tempApp = [FIRApp appNamed:kTempApp];
+  FIRAuth *auth = [FIRAuth authWithApp:tempApp];
+
+  [self.auth fetchProvidersForEmail:emailHint completion:^(NSArray<NSString *> *_Nullable providers,
+                                                           NSError *_Nullable error) {
+    NSString *federatedProviderID = [self federatedAuthProviderFromProviders:providers];
+    // Google case
+    if ([federatedProviderID isEqualToString:FIRGoogleAuthProviderID]) {
+      id<FUIAuthProvider> googleProviderUI;
+      for (id<FUIAuthProvider> provider in self.providers) {
+        if ([provider.providerID isEqualToString:FIRGoogleAuthProviderID]) {
+          googleProviderUI = provider;
+          break;
+        }
+      }
+      [googleProviderUI signOut];
+      [googleProviderUI signInWithDefaultValue:emailHint
+                      presentingViewController:presentingViewController
+                                    completion:^(FIRAuthCredential *_Nullable credential,
+                                                 NSError *_Nullable error,
+                                                 FIRAuthResultCallback  _Nullable result) {
+        if (error) {
+          completion(nil, error);
+          return;
+        }
+
+        [auth signInAndRetrieveDataWithCredential:credential completion:completion];
+      }];
+    }
+  }];
+}
+
+- (nullable NSString *)federatedAuthProviderFromProviders:(NSArray <NSString *> *) providers {
+  NSSet *providerSet =
+      [[NSSet alloc] initWithArray:@[ FIRFacebookAuthProviderID, FIRGoogleAuthProviderID ]];
```

Comment:
nit: use `+[NSSet setWithArray:]`

Also, using `NSSet` for a faster contains when searching an array is not necessarily faster than using `containsObject:` on the array directly, since some cost is incurred when creating the set.

Reply:
Agreed, using the class method but leaving the set, just don't like the idea of creating an array for the sole purpose of look up :).